Documentation/design-docs/first-class-structs.md

```diff
@@ -0,0 +1,608 @@
+            └──▌  indir     ref   
+               └──▌  const(h)  long   0x2425b6229c8 static Fseq[s_dt]
+
+   ▌  stmtExpr  void  (top level) (IL 0x00E...  ???)
+   └──▌  return    void  
+```
+And here is the resulting code, going from a code size of 129 bytes down to 58.
+```
+sub      rsp, 40
+mov      rcx, 0x7FFC8BBB4E10
+mov      edx, 1
+call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
+xor      rax, rax
+mov      rdx, 0x2425B6229C8
+mov      rdx, gword ptr [rdx]
+add      rdx, 8
+vxorpd   ymm0, ymm0
+vmovdqu  qword ptr [rdx], ymm0
+add      rsp, 40
+ret 
+```
+
+Work Items
+----------
+This is a preliminary breakdown of the work into somewhat separable tasks. Those whose descriptions
+are prefaced by '*' have been prototyped in an earlier version of the JIT, and that work is now
+being re-integrated and tested, but may require some cleanup and/or phasing with other work items
+before a PR is submitted.
+
+  ### Mostly-Independent work items
+1.	*Replace block ops with assignments & new nodes.
+
+2.	*Add new fixed-size types, and normalize them in the importer (might be best to do this with or after #1, but not really dependent)
+
+3.	LSRA
+    * Enable support for multiple destination regs, for multi-reg call nodes (for x64/ux)
+    * Handle multiple destination regs for ldp (could be done before or concurrently with the above)
+
+4.	Arm64 & x64/ux: aggressively promote lclVar struct incoming or outgoing args with two 8-byte fields
+
+5.	X64/ux:
+    * modify the handling of multireg struct args to use GT_LIST of GT_IND
+    * remove the restriction to NOT promote things that are multi-reg args, as long as they match (i.e. two 8-byte fields).
+      Pass those using GT_LIST of GT_LCL_VAR.
+    * stop adding extra lclVar copies
+
+6.	Arm64:
+    * Promote 16-byte struct lclVars that are incoming or outgoing args only if they have 2 8-byte fields.
+      Pass those using GT_LIST of GT_LCL_VAR (as above for x64/ux).
+      Note that, if the types do not match, e.g. a TYP_DOUBLE field that will be passed in an integer register,
+      it will require special handling in Lowering and LSRA, as is currently done in the TYP_SIMD8 case.
+    * For other cases, pass as GT_LIST of GT_IND
+    
+    ### Dependent work items:
+    
+7.	*(Depends on 1 & 2): Fully enregister TYP_STRUCT[1|2|3|4|8] with no field accesses.
+
+8.  *(Depends on 1 & 2): Enable value numbering and assertion propagation for struct types.
+
+9.	(Depends on 1 & 2, mostly to avoid conflicts): Add support in prolog to extract fields, and
+    remove the restriction of not promoting incoming reg structs that have more than one field.
+    Note that SIMD types are already reassembled in the prolog.
+    
+10.	(Not really dependent, but probably best done after 1, 2, 5, 6): Add support for assembling
+    non-matching fields into registers for call args and returns. This includes producing the
+    appropriate IR, which may be simply be shifts and or's of the appropriate fields.
```

Comment:
Yes, I think that we should consider doing this.  If the mismatch struct is just passed along it will be done using a register to register move.  For cases where the small fields are accessed and a first cut we could force home the struct into the stack frame.  And ultimately write the code to extract the smaller field from the register.